text/0000-model-data.md

```diff
@@ -0,0 +1,530 @@
+for record data storage and manipulation allowing us to expose a public api for addon authors to experiment with. 
+
+### Motivation
+
+Externally, addons can customize how apps communicate with the server by implementing the Adapter/Serializer APIs but changing how ED deals with relationships, attribute buckets, rollbacks, dirtyness
+and similar issues is extremely challenging and impossible without extremely internal hacks. One can look at popular addons like EmberDataModelFragments
+and see how many private APIs they had to override and hack to implement their funcionality.
+
+Internally, while ED is reasonably well factored between data coming into the system through
+Adapter/Serializers/IdentityMap/Store and data going out through DS.Model/Snapshots/Adapters/Serializers
+, internal handling of the data including relationships and attributes has extremely fuzzy and unclear boundaries.
+
+Data currently lives in internalModels, relationship state objects, computed property caches, relationship
+payload caches, etc.
+
+#### before
+ ![image](https://user-images.githubusercontent.com/715175/33340994-6380c66a-d432-11e7-9f00-ed905e78915a.png)
+
+This RFC proposes rationalizing and extracting ED's core model data handling layer into a ModelData class.
+
+#### after
+![image](https://user-images.githubusercontent.com/715175/33341155-e5f170c2-d432-11e7-9c50-4a3e977331fe.png)
+
+This will allow us to rationalize internal ED APIs, establish clearer internal boundaries, 
+allow experimentation by addon authors, and create a path for internal ED experimentation.
+
+You can think of Model Data as a layer that can receive JSON api payloads for a record,
+apply local changes to it, and can be queried for the current state of the data.
+
+Examples of things this would enable:
+
+1) By shipping a custom ModelData, EmberDataModelFragments can implement a large part of their 
+funcionality without relying on private apis. Spike at [model fragments](https://github.com/igorT/ember-data.model-fragments/tree/igor/model-data)
+
+2) A spike of Ember Data backed by Orbit, can be implemented as an addon, where most of the work
+is in implementing a Model Data backed by Orbit. Spike at [data-orbit](https://github.com/igorT/data-orbit/tree/orbit-model-data)
+
+3) By using an ES6 class for Model Data implementation, this brings us closer to an Emberless 
+Ember Data running.
+
+4) If you needed to implement a GraphQL like projection API, Adapters and Serializers would be enough
+for the loading data, but currently there is no good place to handle client side data interactions.
+ModelData would make it much easier to have a GraphQL ED addon
+
+5) Certain apps and models have a large amount of read only data, which is currently very performance heavy
+to implement in ED. They could use a read only fast model data addon, which would enable a large perf win.
+
+6) Experimenting with schemaless approaches is currently very hard in ED, because internal
```

Comment:
Would love to hear how this would affect https://github.com/hjdivad/ember-m3

Reply:
One of the effects of this API codification is to enable addons like ember-m3, model-fragments &c. to avoid private API, and be easier to maintain between ember-data upgrades.